### DISCUSSION TITLE
Add missing java classes for promesa

### DISCUSSION THREAD

**Discussion by dukefong8** [babashka/babashka#1563] [Q&A]
**Is your feature request related to a problem? Please describe.**
I believe babashka aims to support promesa (though not list officially but doc mentioned a usage with babashka.process). It'd be great to provide full support because promesa.exec.csp has a more robust error handling and non-blocking api (promise based) comparing to core.async (callback based).

**Describe the solution you'd like**
Add all missing java classes from
https://github.com/funcool/promesa/blob/f6cab38000533587afe170e740627cb374f6cdee/src/promesa/exec.cljc#L26. CompletionStage and CountDownLatch, maybe more are needed (require all promesa.* namespaces in test to discover all of them).

**Describe alternatives you've considered**
None

**Additional context**
![image](https://github.com/babashka/babashka/assets/13973254/a33fea67-c0b2-4360-8321-c3932e95b09f)

**Comment by borkdude** [comment]
I'd be happy to add this class, but the next problem will be:

https://github.com/funcool/promesa/blob/f6cab38000533587afe170e740627cb374f6cdee/src/promesa/util.cljc#L31-L34

SCI does not support creating deftypes that implement Java interfaces (this isn't possible without compilation).

So I think promesa doesn't work in this state, but will only work when it's rewritten in a way that works with SCI, currently.

**Comment by dukefong8** [comment]
aha, that's a bummer :-)

Just another thought. It appears that Rich/Alex are working on native java functional interfaces in clojure.core 1.12. Can SCI work with the design they are currently considering? Maybe you should provide some inputs there.

Thanks!

**Reply by borkdude** [reply]
Do you have an example of what they are working on?

**Comment by dukefong8** [comment]
They only showed a DM spreadsheet in https://download.clojure.org/presentations/DesignInPractice.pdf

![image](https://github.com/babashka/babashka/assets/13973254/e5ae3f60-ec43-4141-bc2b-7e80b2911738)

**Reply by borkdude** [reply]
I think this is only about syntax and convenience but there's nothing that fundamentally changes about the `deftype` + Java interfaces problem. If Clojure adopts a more convenient syntax for producing Java function interfaces objects, then I'll try to get that into SCI as well.
